pkg/component/extensions/operatingsystemconfig/operatingsystemconfig.go

```diff
@@ -220,23 +230,174 @@ func (o *operatingSystemConfig) reconcile(ctx context.Context, reconcileFn func(
 		return err
 	}
 
-	fns := o.forEachWorkerPoolAndPurposeTaskFn(func(_ context.Context, osc *extensionsv1alpha1.OperatingSystemConfig, worker gardencorev1beta1.Worker, purpose extensionsv1alpha1.OperatingSystemConfigPurpose) error {
-		d, err := o.newDeployer(osc, worker, purpose)
+	if err := o.updateVersioningSecret(ctx); err != nil {
+		return err
+	}
+
+	fns, err := o.forEachWorkerPoolAndPurposeTaskFn(func(_ context.Context, version int, osc *extensionsv1alpha1.OperatingSystemConfig, worker gardencorev1beta1.Worker, purpose extensionsv1alpha1.OperatingSystemConfigPurpose) error {
+		d, err := o.newDeployer(version, osc, worker, purpose)
 		if err != nil {
 			return err
 		}
 
 		return reconcileFn(d)
 	})
+	if err != nil {
+		return err
+	}
 
 	return flow.Parallel(fns...)(ctx)
 }
 
+type poolHash struct {
+	Pools []poolHashEntry
+}
+
+type poolHashEntry struct {
+	Name           string
+	CurrentVersion int
+	Hashes         map[int]string
+}
+
+func decodePoolHashes(secret *corev1.Secret) (map[string]poolHashEntry, error) {
+	var pools poolHash
+
+	versions := secret.Data[poolHashesSecretField]
+	if len(versions) > 0 {
+		dec := yaml.NewDecoder(bytes.NewReader(versions))
+		if err := dec.Decode(&pools); err != nil {
+			return nil, err
+		}
+	}
+
+	hashesByName := make(map[string]poolHashEntry)
+	for _, entry := range pools.Pools {
+		hashesByName[entry.Name] = entry
+	}
+
+	return hashesByName, nil
+}
+
+func encodePoolHashes(pools *poolHash, secret *corev1.Secret) error {
+	var buf bytes.Buffer
+	enc := yaml.NewEncoder(&buf)
+	if err := enc.Encode(&pools); err != nil {
+		return err
+	}
+
+	if secret.Data == nil {
+		secret.Data = map[string][]byte{}
+	}
+	secret.Data[poolHashesSecretField] = buf.Bytes()
+	return nil
+}
+
+func (o *operatingSystemConfig) updateVersioningSecret(ctx context.Context) error {
+	// FIXME set controller etc.
```

Comment:
?

Reply:
Oh, I forgot to remove that comment. I was wondering whether it might make sense to set some controllerReference for the secret.